### PR TITLE
[Process] Pass the commandline as array to `proc_open()`

### LINKED ISSUE #43162: [Process] Allow running multiple commands at once
**Description**
When running a lot of small shell commands after each other, the overhead of creating the underlying processes becomes noticeable. It would be handy if the process component would support chaining commands.

Working directly on a Linux/Windows shell you would typically do this:
```shell
# Linux
$ foo; bar; baz
$ foo && bar && baz

# Windows
$ foo & bar & baz
$ foo && bar && baz
```
(Yes there are other chaining operators as well like `||`.)

This is currently not possible with the process component.

**API Example**
I like that `Symfony\Component\Process\Process` became somewhat immutable now (the command line is baked in after the constructor ran), so the API could maybe look like this:

```php
$process = new Process(['foo']);

// foo; bar
$process = $process->withChainedCommand(['bar']); // returns a new process instance

// foo && bar
$process = $process->withAndChainedCommand(['bar']); // returns a new process instance
```

What do you think? Is this a worthwhile addition?

### PR DESCRIPTION
**Pull request by ausi**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix https://github.com/symfony/symfony/issues/43162#issuecomment-931145776
| License       | MIT

Since PHP 7.4, `proc_open()` accepts an array which makes it twice as fast (4x as fast with PHP 8.3), see https://github.com/symfony/symfony/issues/43162#issuecomment-1787670224

This pull request enables calling `proc_open()` with an array if:

1. `$this->commandline` is an array
2. and we are not on Windows
3. and PHP was compiled without `--enable-sigchild`

### CODE REVIEW

File: src/Symfony/Component/Process/Process.php [hunk 1]
```diff
@@ -327,6 +322,11 @@ public function start(callable $callback = null, array $env = [])
             // last exit code is output on the fourth pipe and caught to work around --enable-sigchild
             $descriptors[3] = ['pipe', 'w'];
 
+            if (\is_array($commandline)) {
+                // exec is mandatory to deal with sending a signal to the process
+                $commandline = 'exec '.$this->buildShellCommandline($commandline);
+            }
```
Comment (ausi) [inline src/Symfony/Component/Process/Process.php:324]: I’m not sure if `exec ` is needed here, on my Platform (macOS) it does not make a difference.

**Comment by ausi** [comment]
The failing unit test in PHP 8.3 seems to be a bug in PHP: https://github.com/php/php-src/issues/12589

**Comment by OskarStark** [comment]
It is a new feature and must therefore target 7.1 because of feature freeze period

Should we wait wait for a PHP bugfix before merge?

**Comment by ausi** [comment]
> It is a new feature and must therefore target 7.1 because of feature freeze period

Changed the base to 7.0 (there is no 7.1 branch yet)

> Should we wait wait for a PHP bugfix before merge?

I think so, yes.

**Comment by ausi** [comment]
> > Should we wait wait for a PHP bugfix before merge?
> 
> I think so, yes.

The PHP bug got closed, the new behavior is intended. I updated the pull request to handle that case.

File: src/Symfony/Component/HttpClient/Tests/HttpClientTestCase.php [hunk 2]
```diff
@@ -327,7 +328,12 @@ private static function startVulcain(HttpClientInterface $client)
             'KEY_FILE' => __DIR__.'/Fixtures/tls/server.key',
             'CERT_FILE' => __DIR__.'/Fixtures/tls/server.crt',
         ]);
-        $process->start();
+
+        try {
+            $process->start();
+        } catch (RuntimeException $e) {
+            self::markTestSkipped('vulcain failed: '.$e->getMessage());
+        }
```
Comment (derrabus) [inline src/Symfony/Component/HttpClient/Tests/HttpClientTestCase.php:335]: Unrelated change?

File: src/Symfony/Component/Process/Process.php [hunk 3]
```diff
@@ -340,10 +340,17 @@ public function start(callable $callback = null, array $env = []): void
...
         }
 
+        $lastError = 'unknown reason';
+        set_error_handler(function ($type, $msg) use (&$lastError) {
+            $lastError = $msg;
+
+            return true;
+        });
         $process = @proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $envPairs, $this->options);
+        restore_error_handler();
```
Comment (derrabus) [inline src/Symfony/Component/Process/Process.php]:
```suggestion
        try {
            $process = @proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $envPairs, $this->options);
        } finally {
            restore_error_handler();
        }
```

File: src/Symfony/Component/HttpClient/Tests/HttpClientTestCase.php (hunk 2 above)
Comment (ausi) [reply src/Symfony/Component/HttpClient/Tests/HttpClientTestCase.php:335]: No, by using `proc_open()` with an array, starting the process can fail directly since PHP 8.3 if the binary does not exist. This change is needed to restore the desired behavior of that test.

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 4]
```diff
@@ -66,6 +66,24 @@ public function testInvalidCwd()
         $cmd->run();
     }
 
+    public function testInvalidCommand()
+    {
+        $cmds = [
+            new Process(['invalid']),
+            Process::fromShellCommandline('invalid'),
+        ];
```
Comment (OskarStark) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]: What about using a data provider?

Comment (stof) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: This would indeed be better with a data provider or 2 tests rather than a loop inside the test emulating having multiple tests

Comment (ausi) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Done in d603793013e2294d8f97c378b0256e86c8dfedb9.

**Comment by ausi** [comment]
~~The PHP bug https://github.com/php/php-src/issues/12589 got reopened, so we should wait until that was resolved there I think.~~ 
UPDATE 2023-11-13: was closed again

**Comment by ausi** [comment]
@carsonbot the 7.1 branch does not exist yet.

File: src/Symfony/Component/Process/Process.php [hunk 5]
```diff
@@ -338,10 +338,20 @@ public function start(callable $callback = null, array $env = []): void
...
+        });
+        try {
+            $process = @proc_open($commandline, $descriptors, $this->processPipes->pipes, $this->cwd, $envPairs, $this->options);
+        } finally {
+            restore_error_handler();
+        }
 
         if (!\is_resource($process)) {
-            throw new RuntimeException('Unable to launch a new process.');
+            throw new RuntimeException('Unable to launch a new process: '.$lastError);
```
Comment (ausi) [inline src/Symfony/Component/Process/Process.php]: Should we throw a `ProcessFailedException` here? That would improve backwards compatibility I think.

Or should we maybe create a new exception like `ProcessStartFailedException`?

File: src/Symfony/Component/Process/Process.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Process.php:324]: I think so, let's keep it.

File: src/Symfony/Component/Process/Process.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Process.php]: ProcessFailedException would be fine to me

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for working on this!
Can't we use an array on Windows also? Sorry I didn't read everything :grimacing: 
While at it, WDYT of the warnings on appveyor :)
```
'where' is not recognized as an internal or external command,
operable program or batch file.
'where' is not recognized as an internal or external command,
operable program or batch file.
ERROR: Invalid argument or option - '/usr/local/php/bin/php-invalid'.
Type "WHERE /?" for usage help.
ERROR: Invalid argument or option - '/usr/local/php/bin/php-invalid'.
Type "WHERE /?" for usage help.
```

File: src/Symfony/Component/Process/Process.php (hunk 5 above)
Comment (ausi) [reply src/Symfony/Component/Process/Process.php]: I created a new `ProcessStartFailedException` that extends the `ProcessFailedException` in 833bf1d2dbd88559d84d8528aa825c45df1bb86b.
This is a clean API I think and makes the changed behavior in PHP 8.3 a little bit more BC.

**Comment by nicolas-grekas** [comment]
Thank you @ausi.

**Comment by ausi** [comment]
> Can't we use an array on Windows also?

I don’t think so as we use some heavy `cmd.exe` magic on Windows. Also, there would be no real difference on Windows I think as internally PHP just converts the array back into a string and sets `bypass_shell = 1` which we also set.
